File: qc-core/src/main/java/com/eager/questioncloud/core/domain/social/NaverAPI.java
package com.eager.questioncloud.core.domain.social;

import com.eager.questioncloud.core.domain.user.vo.AccountType;
import com.eager.questioncloud.core.exception.CustomException;
import com.eager.questioncloud.core.exception.Error;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

@Component
public class NaverAPI extends SocialAPI {
    @Value("${NAVER_CLIENT_ID}")
    private String NAVER_CLIENT_ID;

    @Value("${NAVER_CLIENT_SECRET}")
    private String NAVER_CLIENT_SECRET;

    @Override
    public String getAccessToken(String code) {
        WebClient webClient = WebClient.create("https://nid.naver.com/oauth2.0/token");

        SocialAccessToken res = webClient.get()
            .uri(uriBuilder -> uriBuilder
                .queryParam("grant_type", "authorization_code")
                .queryParam("client_id", NAVER_CLIENT_ID)
                .queryParam("client_secret", NAVER_CLIENT_SECRET)
                .queryParam("code", code)
                .build())
            .exchangeToMono(response -> response.bodyToMono(SocialAccessToken.class))
            .block();

        if (res == null || res.access_token() == null) {
            throw new CustomException(Error.FAIL_SOCIAL_LOGIN);
        }

        return res.access_token();
    }

    @Override
    public SocialUserInfo getUserInfo(String accessToken) {
        WebClient webClient = WebClient.create("https://openapi.naver.com/v1/nid/me");

        NaverUserInfoAPIResponse apiResponse = webClient.get()
            .headers(header -> {
                header.add("Authorization", "Bearer " + accessToken);
            })
            .exchangeToMono(response -> response.bodyToMono(NaverUserInfoAPIResponse.class))
            .block();

        if (apiResponse == null || apiResponse.response() == null || apiResponse.response().id() == null) {
            throw new CustomException(Error.FAIL_SOCIAL_LOGIN);
        }

        NaverUserInfo naverUserInfo = apiResponse.response();

        return new SocialUserInfo(naverUserInfo.id(), naverUserInfo.email(), naverUserInfo.nickname(), AccountType.NAVER);
    }

    @Override
    SocialPlatform getSocialPlatform() {
        return SocialPlatform.NAVER;
    }

    record NaverUserInfoAPIResponse(NaverUserInfo response) {
    }

    record NaverUserInfo(String id, String email, String nickname, String mobile, String profile_image) {
    }
}

File: qc-core/qc-user/qc-user-social/src/main/kotlin/com/eager/questioncloud/social/naver/NaverAPI.kt
package com.eager.questioncloud.social.naver

import com.eager.questioncloud.common.http.HttpClient
import com.eager.questioncloud.common.http.HttpRequest
import com.eager.questioncloud.social.SocialAPI
import com.eager.questioncloud.social.SocialAccessToken
import com.eager.questioncloud.social.SocialPlatform
import com.eager.questioncloud.social.SocialUserInfo
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component
class NaverAPI(
    private val objectMapper: ObjectMapper,
    private val httpClient: HttpClient,
) : SocialAPI {
    @Value("\${NAVER_CLIENT_ID}")
    private lateinit var NAVER_CLIENT_ID: String
    
    @Value("\${NAVER_CLIENT_SECRET}")
    private lateinit var NAVER_CLIENT_SECRET: String
    
    override fun getAccessToken(code: String): String {
        val query = mutableMapOf<String, String>()
        query["grant_type"] = "authorization_code"
        query["client_id"] = NAVER_CLIENT_ID
        query["client_secret"] = NAVER_CLIENT_SECRET
        query["code"] = code
        
        val request = HttpRequest(url = "https://nid.naver.com/oauth2.0/token", query = query)
        val response = httpClient.get(request, SocialAccessToken::class.java)
        return response.access_token
    }
    
    override fun getUserInfo(accessToken: String): SocialUserInfo {
        val headers = mapOf("Authorization" to "Bearer $accessToken")
        val request = HttpRequest("https://openapi.naver.com/v1/nid/me", headers = headers)
        val response = httpClient.get(request, NaverUserInfoAPIResponse::class.java).response
        
        return SocialUserInfo(
            response.id,
            response.email,
            response.nickname,
            SocialPlatform.NAVER
        )
    }
    
    override fun getSocialPlatform(): SocialPlatform {
        return SocialPlatform.NAVER
    }
}

private data class NaverUserInfoAPIResponse(val response: NaverUserInfo)

private data class NaverUserInfo(
    val id: String,
    val email: String,
    val nickname: String,
    val mobile: String,
    val profile_image: String
)
